src/base58.cpp

```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongs(const Span<const unsigned char>& input, int leadingZeros) -> std::vector<int64_t> {
+    int effectiveLength = input.size() - leadingZeros;
+    std::vector<int64_t> inputAsLongs((effectiveLength + (GROUP_SIZE - 1)) / GROUP_SIZE);
+    int groupOffset = ((GROUP_SIZE - (effectiveLength % GROUP_SIZE)) % GROUP_SIZE) - leadingZeros;
+
+    for (size_t i = leadingZeros; i < input.size(); ++i) {
+        int index = (groupOffset + i) / GROUP_SIZE;
+        inputAsLongs[index] <<= 8;
+        inputAsLongs[index] |= input[i];
+    }
+
+    return inputAsLongs;
+}
+
 std::string EncodeBase58(Span<const unsigned char> input)
 {
-    // Skip & count leading zeroes.
-    int zeroes = 0;
-    int length = 0;
-    while (input.size() > 0 && input[0] == 0) {
-        input = input.subspan(1);
-        zeroes++;
+    auto leadingZeros{0U};
+    while (leadingZeros < input.size() && input[leadingZeros] == 0) {
+        ++leadingZeros;
     }
-    // Allocate enough space in big-endian base58 representation.
-    int size = input.size() * 138 / 100 + 1; // log(256) / log(58), rounded up.
-    std::vector<unsigned char> b58(size);
-    // Process the bytes.
-    while (input.size() > 0) {
-        int carry = input[0];
-        int i = 0;
-        // Apply "b58 = b58 * 256 + ch".
-        for (std::vector<unsigned char>::reverse_iterator it = b58.rbegin(); (carry != 0 || i < length) && (it != b58.rend()); it++, i++) {
-            carry += 256 * (*it);
-            *it = carry % 58;
-            carry /= 58;
-        }
 
-        assert(carry == 0);
-        length = i;
-        input = input.subspan(1);
+    auto size = 1 + input.size() * LOG_256_58_RATIO / BASE_SCALE;
+    std::string result;
+    result.reserve(leadingZeros + size);
+    result.assign(leadingZeros, '1');
+
+    auto inputAsLongs = ConvertToLongs(input, leadingZeros);
+    for (auto i{0U}; i < inputAsLongs.size();) {
```

Comment:
This construct with the two nested loops is not evident. Maybe some comment would be useful. The inner loop does not depend on the value of `i`. Is it repeated for as long as all `inputAsLongs` elements are reduced to 0?

Reply:
Fantastic comment, the internal loop should skip the leading zeros, indeed.